go/engine/selfprovision.go

```diff
@@ -0,0 +1,295 @@
+	// Pass the UV here so the passphrase stream is cached on the provisional
+	// login context
+	m = m.WithNewProvisionalLoginContextForUserVersionAndUsername(uvOld, e.G().Env.GetUsername())
+
+	// From this point on, if there's an error, we abort
+	// the transaction.
+	defer func() {
+		if tx != nil {
+			tx.Abort()
+		}
+		if err == nil {
+			m = m.CommitProvisionalLogin()
+		} else if e.User != nil {
+			m.CDebugf("Error in self provision, reverting to original active device: %v", err)
+			m = m.WithGlobalActiveDevice()
+			salt, err := e.User.GetSalt()
+			if err != nil {
+				m.CDebugf("unable to GetSalt: %v", err)
+				return
+			}
+			// Atomically swap to the new config and active device
+			if err := m.SwitchUserNewConfigActiveDevice(uvOld, e.User.GetNormalizedName(), salt,
+				deviceIDOld, sigKeyOld, encKeyOld, deviceNameOld); err != nil {
+				return
+			}
+		}
+	}()
+
+	// run the LoginLoadUser sub-engine to load a user
+	ueng := newLoginLoadUser(e.G(), e.G().Env.GetUsername().String())
+	if err = RunEngine2(m, ueng); err != nil {
+		return err
+	}
+	e.User = ueng.User()
+
+	activeDevice := e.G().ActiveDevice
+	encKey, err := activeDevice.EncryptionKey()
+	if err != nil {
+		return err
+	}
+	sigKey, err := activeDevice.SigningKey()
+	if err != nil {
+		return err
+	}
+	keys := libkb.NewDeviceWithKeysOnly(sigKey, encKey)
+	if _, err := keys.Populate(m); err != nil {
+		return err
+	}
+
+	e.perUserKeyring, err = libkb.NewPerUserKeyring(e.G(), e.User.GetUID())
+	if err != nil {
+		return err
+	}
+	// Make new device keys and sign them with current device keys
+	if err = e.provision(m, keys); err != nil {
+		return err
+	}
+	// commit the config changes
+	if err := tx.Commit(); err != nil {
+		return err
+	}
+	// Zero out the TX so that we don't abort it in the defer()
+	// exit.
+	tx = nil
+
+	verifyLocalStorage(m, e.User.GetNormalizedName().String(), e.User.GetUID())
+
+	e.clearCaches(m)
+	e.sendNotification()
+	return nil
+}
+
+func (e *SelfProvisionEngine) provision(m libkb.MetaContext, keys *libkb.DeviceWithKeys) error {
+	// After obtaining login session, this will be called before the login
+	// state is released.  It signs this new device with the current device.
+	u := e.User
+	uv := u.ToUserVersion()
+
+	// Set the active device to be a special provisional key active device,
+	// which keeps a cached copy around for DeviceKeyGen, which requires it to
+	// be in memory.  It also will establish a NIST so that API calls can
+	// proceed on behalf of the user.
+	m = m.WithProvisioningKeyActiveDevice(keys, uv)
+
+	// need lksec to store device keys locally
+	if err := e.fetchLKS(m, keys.EncryptionKey()); err != nil {
+		return err
+	}
+	if err := e.makeDeviceKeysWithSigner(m, keys.SigningKey()); err != nil {
+		return err
+	}
+
+	// now store the secrets for our new device
+	encKey, err := m.ActiveDevice().EncryptionKey()
+	if err != nil {
+		return err
+	}
+	if err := e.fetchLKS(m, encKey); err != nil {
```

Comment:
Is it possible to update `TestPassphraseRecover` in systests to see if this stuff is actually working?  The flow would be to make a device, make a paper key, clone the device, self-provision, and then check that passphrase recover works on the self-provisioned device.

Reply:
Otherwise, the LKSec stuff is pretty easy to get wrong.  I worry that, for instance, the LKSec system hasn't updated for the new device, and is some mishmash of the old device and the new.

Reply:
sounds good!